File: Java/Submission3/app/src/main/java/com/makhalibagas/submission2/Provider/MyContentProvider.java
package com.makhalibagas.submission2.Provider;

import android.content.ContentProvider;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.net.Uri;

import com.makhalibagas.submission2.Database.UserHelper;

import java.util.Objects;

import static com.makhalibagas.submission2.Database.UserDbContract.AUTHORITY;
import static com.makhalibagas.submission2.Database.UserDbContract.UserColumns.CONTENT_URI;
import static com.makhalibagas.submission2.Database.UserDbContract.UserColumns.TABLE_USER_NAME;

public class MyContentProvider extends ContentProvider {
    private static final int USER = 0;
    private static final int USER_ID = 1;
    UserHelper userHelper;


    private static final UriMatcher uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);

    static {

        uriMatcher.addURI(AUTHORITY,TABLE_USER_NAME,USER);
        uriMatcher.addURI(AUTHORITY,TABLE_USER_NAME + "/#",USER_ID);
    }
    @Override
    public boolean onCreate() {
        // TODO: Implement this to initialize your content provider on startup.
        userHelper = UserHelper.getInstance(getContext());
        userHelper.open();
        return true;
    }

    @Override
    public Cursor query(Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        Cursor cursor;
        switch (uriMatcher.match(uri)){
            case USER:
                cursor = userHelper.queryAll();
                break;
            case USER_ID:
                cursor = userHelper.queryById(uri.getLastPathSegment());
                break;
            default:
                cursor = null;
                break;
        }

        if (cursor != null){
            cursor.setNotificationUri(Objects.requireNonNull(getContext()).getContentResolver(),uri);
        }
        return cursor;
        // TODO: Implement this to handle query requests from clients.
    }

    @Override
    public String getType(Uri uri) {
        // TODO: Implement this to handle requests for the MIME type of the data
        // at the given URI.
        return null;
    }

    @Override
    public Uri insert(Uri uri, ContentValues values) {
        long added;
        Uri contentUri = null;
        switch (uriMatcher.match(uri)){
            case USER:
                added = userHelper.InsertProvider(values);
                if (added > 0) {
                    contentUri = ContentUris.withAppendedId(CONTENT_URI, added);
                }
                break;
            default:
                added = 0;
                break;

        }


        if (added > 0) {
            Objects.requireNonNull(getContext()).getContentResolver().notifyChange(uri, null);
        }
        return contentUri;
    }


    @Override
    public int update(Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        int update;
        switch (uriMatcher.match(uri)){
            case USER_ID:
                update = userHelper.UpdateProvider(uri.getLastPathSegment(),values);
                break;
            default:
                update = 0;
                break;
        }


        if (update > 0) {
            Objects.requireNonNull(getContext()).getContentResolver().notifyChange(uri, null);
        }
        return update;
    }


    @Override
    public int delete(Uri uri, String selection, String[] selectionArgs) {
        // Implement this to handle requests to delete one or more rows.
        int delete;
        switch (uriMatcher.match(uri)){
            case USER_ID:
                delete = userHelper.DeleteProvider(uri.getLastPathSegment());
                break;
            default:
                delete = 0;
                break;
        }
        if (delete > 0) {
            Objects.requireNonNull(getContext()).getContentResolver().notifyChange(uri, null);

        }
        return delete;
    }

}

File: Kotlin/Submission3/app/src/main/java/com/makhalibagas/githubuser/provider/MyContentProvider.kt
package com.makhalibagas.githubuser.provider

import android.content.ContentProvider
import android.content.ContentValues
import android.content.Context
import android.content.UriMatcher
import android.database.Cursor
import android.net.Uri
import com.makhalibagas.githubuser.database.UserContract.AUTHOR_USER
import com.makhalibagas.githubuser.database.UserContract.UserColumns.Companion.CONTENT_URI_USER
import com.makhalibagas.githubuser.database.UserContract.UserColumns.Companion.TABLE_USER
import com.makhalibagas.githubuser.database.UserHelper

class MyContentProvider : ContentProvider() {


    companion object{
        private const val USER = 1
        private const val USER_ID = 2
        private lateinit var userHelper: UserHelper

        private val uriMatcher = UriMatcher(UriMatcher.NO_MATCH)

        init {
            uriMatcher.addURI(AUTHOR_USER, TABLE_USER, USER)
            uriMatcher.addURI(AUTHOR_USER, "$TABLE_USER/#", USER_ID)
        }
    }
    override fun onCreate(): Boolean {
        userHelper = UserHelper.getDatabase(context as Context)
        userHelper.open()
        return true

    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {

        var cursor : Cursor? = null
        when(uriMatcher.match(uri)){
            USER -> cursor = userHelper.queryAll()
            USER_ID -> userHelper.queryById(uri.lastPathSegment.toString())

            else -> cursor = null
        }

        return cursor

    }


    override fun getType(uri: Uri): String? {
        return null
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        val added : Long = when(USER){
            uriMatcher.match(uri) -> userHelper.insert(values)
            else -> 0
        }

        context?.contentResolver?.notifyChange(CONTENT_URI_USER, null)
        return Uri.parse("$CONTENT_URI_USER/$added")
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        val update : Int = when(USER_ID){
            uriMatcher.match(uri) -> userHelper.update(uri.lastPathSegment.toString(), values)
            else -> 0
        }

        context?.contentResolver?.notifyChange(CONTENT_URI_USER, null)

        return update
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        val delete = when(USER_ID){
            uriMatcher.match(uri) -> userHelper.delete(uri.lastPathSegment.toString())
            else -> 0
        }

        context?.contentResolver?.notifyChange(CONTENT_URI_USER, null)
        return delete
    }
}
